subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

```diff
@@ -154,6 +154,114 @@ The following method on `Configuration` is deprecated for removal:
 
 Obtain the set of all configurations from the project's `configurations` container instead.
 
+[[test_framework_implementation_dependencies]]
+==== Relying on automatic test framework implementation dependencies
+
+In some cases, Gradle will load JVM test framework dependencies from the Gradle distribution in order to
+execute tests. This existing behavior can lead to test framework dependency version conflicts on the test
+classpath. To avoid these conflicts, this behavior is deprecated and will be removed in Gradle 9.0. Tests
+using TestNG are unaffected.
+
+In order to prepare to for this change in behavior, either declare the required dependencies explicitly,
+or migrate to link:jvm_test_suite_plugin.html[Test Suites], where these dependencies are managed automatically.
+
+===== Test Suites
+
+If you are using test suites, this change will not affect you. The test framework dependencies are managed
+by the suite itself by automatically declaring the proper dependencies on the test suite's configurations.
+In order to minimally migrate to test suites, use the following code:
+
+=====
+[.multi-language-sample]
+======
+.build.gradle.kts
+[source,kotlin]
+----
+testing {
+    suites {
+        named<JvmTestSuite>("test") {
+            // If using JUnit Jupiter
+            useJUnitJupiter()
+
+            // If using JUnit 4
+            useJUnit()
+        }
+    }
+}
+----
+======
+[.multi-language-sample]
+======
+.build.gradle
+[source,groovy]
+----
+testing {
+    suites {
+        test {
+            // If using JUnit Jupiter
+            useJUnitJupiter()
+
+            // If using JUnit 4
+            useJUnit()
+        }
+    }
+}
+----
+======
+=====
+
+===== Manually declaring dependencies
+
+This change in behavior can also be mitigated without migrating to test suites. In this case, the proper
+dependencies must be manually declared on the test runtime classpath:
+
+  * If using JUnit 5, an explicit `runtimeOnly` dependency on `junit-platform-launcher` is required
+  in addition to the existing `implementation` dependency on the test engine.
+  * If using JUnit 4, only the existing `implementation` dependency on `junit` 4 is required.
+  * If using JUnit 3, a test `runtimeOnly` dependency on `junit` 4 is required in addition to a
+  `compileOnly` dependency on `junit` 3.
```

Comment:
❓ So previously, this would've been a implementation dependency on junit3.  What was happening at runtime in that case?  Was the test worker getting an implicit junit4 library at runtime and the junit3 library was being ignored?  I'm trying to understand why junit3 should now be `compileOnly`...